File: src/Enterprise/VB.NET/patterns/Mediator.vb
Imports System
Imports System.Collections.Generic

Friend Module MediatorExample
    Private NotInheritable Class CheckoutMediator
        Private ReadOnly colleagues As New Dictionary(Of String, Action(Of String, String))(StringComparer.Ordinal)

        Public Sub Register(name As String, receiver As Action(Of String, String))
            colleagues(name) = receiver
        End Sub

        Public Function Send(sender As String, recipient As String, message As String) As Boolean
            Dim receiver As Action(Of String, String) = Nothing
            If Not colleagues.TryGetValue(recipient, receiver) Then Return False
            receiver(sender, message)
            Return True
        End Function
    End Class

    Friend Function Run() As Boolean
        Dim deliveries As New List(Of String)
        Dim mediator As New CheckoutMediator

        mediator.Register("payment", Sub(sender, message) deliveries.Add($"payment<-{sender}:{message}"))
        mediator.Register("inventory", Sub(sender, message) deliveries.Add($"inventory<-{sender}:{message}"))

        Dim reserveDelivered = mediator.Send("payment", "inventory", "reserve")
        Dim chargedDelivered = mediator.Send("inventory", "payment", "reserved")
        Dim unknownRejected = Not mediator.Send("payment", "shipping", "dispatch")

        Return reserveDelivered AndAlso
            chargedDelivered AndAlso
            unknownRejected AndAlso
            String.Join(">", deliveries) = "inventory<-payment:reserve>payment<-inventory:reserved"
    End Function
End Module

File: src/Enterprise/C#/patterns/Mediator.cs
using System;
using System.Collections.Generic;

namespace Genkidama.PatternExamples;

public static class MediatorExample
{
    public static bool Run()
    {
        var events = new List<string>();
        var colleagues = new Dictionary<string, Action<string, string>>(StringComparer.Ordinal)
        {
            ["inventory"] = (sender, message) => events.Add($"inventory<-{sender}:{message}"),
            ["payment"] = (sender, message) => events.Add($"payment<-{sender}:{message}")
        };

        void Send(string sender, string recipient, string message)
        {
            if (!colleagues.TryGetValue(recipient, out var receiver))
            {
                throw new InvalidOperationException($"Unknown colleague: {recipient}");
            }

            receiver(sender, message);
        }

        void Payment(string message) => Send("payment", "inventory", message);
        void Inventory(string message) => Send("inventory", "payment", message);

        Payment("paid");
        Inventory("reserved");

        var rejectedUnknown = false;
        try
        {
            Send("payment", "shipping", "paid");
        }
        catch (InvalidOperationException)
        {
            rejectedUnknown = true;
        }

        return string.Join('>', events) ==
               "inventory<-payment:paid>payment<-inventory:reserved" &&
               rejectedUnknown;
    }
}
